File: 10-GrantPasswordTokenUsecase.java
@Service
@AllArgsConstructor
public class GrantPasswordTokenUsecase {
    private final UserRepository userRepository;
    private final UserTokenRepository userTokenRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtFactory jwtFactory;
    private final SqlQueryRepository sqlQueryRepository;
    private final DomainEventPublisher publisher;

    @Transactional
    public Mono<TokensDto> execute(@NonNull Command cmd) {
        return Mono.just(new Context(cmd))
            .flatMap(this::findUserWithPassword)
            .doOnNext(ctx -> ctx.accessToken = jwtFactory.buildAccessToken(ctx.userDetails))
            .flatMap(this::buildRefreshToken)
            .flatMap(this::publishEvent)
            .flatMap(this::setLastLoggedOn)
            .map(Context::toTokens);
    }

    private Mono<Context> findUserWithPassword(Context ctx) {
        return sqlQueryRepository.findUserDetailsByEmail(ctx.cmd.getUsername())
            .filter(user -> passwordEncoder.matches(ctx.cmd.getPassword(), user.getPasswordHash()))
            .doOnNext(user -> ctx.userDetails = user)
            .thenReturn(ctx);
    }

    public Mono<Context> buildRefreshToken(Context ctx) {
        return userTokenRepository.deleteByUserIdAndType(ctx.userDetails.getId(), UserTokenType.REFRESH)
            .thenReturn(new UserToken(ctx.userDetails.getId(), UserTokenType.REFRESH))
            .flatMap(userTokenRepository::save)
            .doOnNext(userToken -> ctx.refreshToken = jwtFactory.buildRefreshToken(userToken.getValue()))
            .thenReturn(ctx);
    }

    public Mono<Context> publishEvent(Context ctx) {
        return publisher.publish(ctx.userDetails,
                new UserLoggedEvent()
                    .setEmail(ctx.userDetails.getEmail())
                    .setIp(ctx.cmd.ip),
                ctx.userDetails.getId())
            .thenReturn(ctx);
    }

    public Mono<Context> setLastLoggedOn(Context ctx) {
        return userRepository.updateLastLoggedOn(ctx.userDetails.getId())
            .thenReturn(ctx);
    }

    @RequiredArgsConstructor
    private static class Context {
        final Command cmd;
        UserDetailsDto userDetails;
        String accessToken;
        String refreshToken;

        TokensDto toTokens() {
            return new TokensDto()
                .setAccessToken(accessToken)
                .setRefreshToken(refreshToken);
        }
    }

    @Schema(name = "LoginUserCommand")
    @Data
    @Accessors(chain = true)
    public static class Command {
        @NotBlank
        @Email
        private String username;

        @NotBlank
        private String password;

        @JsonIgnore
        private String ip;
    }

    @Schema(name = "TokenResponse")
    @Data
    @Accessors(chain = true)
    public static class TokensDto {
        private String accessToken;
        @JsonIgnore
        private String refreshToken;
    }
}

File: 10-GrantPasswordTokenUsecase.kt
@Service
class GrantPasswordTokenUsecase(
    private val userRepository: UserRepository,
    private val userTokenRepository: UserTokenRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtFactory: JwtFactory,
    private val sqlQueryRepository: SqlQueryRepository,
    private val publisher: DomainEventPublisher,
) {
    @Transactional
    suspend fun execute(cmd: Command): TokensDto {
        val userDetails = sqlQueryRepository.findUserDetailsByEmail(cmd.username!!) ?: throw PasswordDontMatchException()
        if (!passwordEncoder.matches(cmd.password, userDetails.passwordHash)) throw PasswordDontMatchException()
        userRepository.updateLastLoggedOn(userDetails.id)
        publisher.publish(userDetails, UserLoggedEvent(userDetails.email, cmd.ip))
        return TokensDto(
            accessToken = jwtFactory.buildAccessToken(userDetails),
            refreshToken = buildRefreshToken(userDetails.id),
        )
    }

    suspend fun buildRefreshToken(userId: UUID): String {
        userTokenRepository.deleteByUserIdAndType(userId, UserTokenType.REFRESH)
        val token = userTokenRepository.save(UserToken(userId, UserTokenType.REFRESH))
        return jwtFactory.buildRefreshToken(token.value)
    }

    @Schema(name = "LoginUserCommand")
    data class Command(
        @field:NotBlank @field:Email var username: String?,
        @field:NotBlank var password: String?,
        @field:JsonIgnore var ip: String?,
    )

    @Schema(name = "TokenResponse")
    data class TokensDto(
        private var accessToken: String,
        @JsonIgnore var refreshToken: String,
    )

    @ResponseStatus(HttpStatus.FORBIDDEN)
    class PasswordDontMatchException : Throwable("Password do not match")
}

